File: 79. Word Search.cpp
class Solution {
private:
    vector<int> diff = {-1, 0, 1, 0, -1};
public:
    bool exist(vector<vector<char>>& board, string word) {
        if (board.empty() || board[0].empty() || word.empty())
            return false;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == word[0]) {
                    if (DFS(board, i, j, word, 0))
                        return true;
                }
            }
        }
        return false;
    }

    bool DFS(vector<vector<char>>& board, int i, int j, string &word, int p) {
        if (p == word.size())
            return true;
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '\0' || board[i][j] != word[p])
            return false;
        int tmp = board[i][j];
        board[i][j] = '\0';
        for (int k = 0; k < 4; k++) {
            if (DFS(board, i + diff[k], j + diff[k + 1], word, p + 1))
                return true;
        }
        board[i][j] = tmp;
        return false;
    }
};

File: java/79. Word Search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (word.length() == 0 || board.length == 0 || board[0].length == 0) {
            return false;
        }
        for (int i = 0; i < board.length; ++i) {
            for (int j = 0; j < board[0].length; ++j) {
                if (board[i][j] == word.charAt(0)) {
                    HashSet<Integer> visited = new HashSet<>();
                    visited.add(i * board[0].length + j);
                    if (word.length() == 1 || dfs(board, i - 1, j, word, 1, visited) || dfs(board, i + 1, j, word, 1, visited) || dfs(board, i, j - 1, word, 1, visited) || dfs(board, i, j + 1, word, 1, visited)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, int i, int j, String word, int idx, HashSet<Integer> visited) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || visited.contains(i * board[0].length + j) || board[i][j] != word.charAt(idx)) {
            return false;
        }
        if (idx == word.length() - 1) {
            return true;
        }
        visited.add(i * board[0].length + j);
        if (dfs(board, i - 1, j, word, idx + 1, visited) || dfs(board, i + 1, j, word, idx + 1, visited) || dfs(board, i, j - 1, word, idx + 1, visited) || dfs(board, i, j + 1, word, idx + 1, visited)) {
            return true;
        }
        visited.remove(i * board[0].length + j);
        return false;
    }
}
